### PR TITLE
Add normalization with statistics to many statistics preprocessors

### LINKED ISSUE #2185: New preprocessor: subtract global mean from field
A common operation in climate science is to subtract global means from a field. I will implement that.

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR adds the argument `normalize` to all statistical preprocessors that are based on `Cube.collapsed` (exceptions: multi-model statistics and ensemble statistics [their call signature is different], and climate statistics [we already have the preprocessor `anomaly`, which does the very same as this PR]). This PR skips statistical preprocessors that use `Cube.aggregated_by` since these need extra care and cannot simply make use of iris' rich broadcasting capabilities.

For example, this is useful to subtract the global mean from a field or divide by the global standard deviation:

```yaml
area_statistics:
  operator: mean
  normalize: subtract
```

Closes #2185

Link to documentation:
- https://esmvaltool--2189.org.readthedocs.build/projects/ESMValCore/en/2189/recipe/preprocessor.html#zonal-statistics
- https://esmvaltool--2189.org.readthedocs.build/projects/ESMValCore/en/2189/recipe/preprocessor.html#meridional-statistics
- https://esmvaltool--2189.org.readthedocs.build/projects/ESMValCore/en/2189/recipe/preprocessor.html#area-statistics
- https://esmvaltool--2189.org.readthedocs.build/projects/ESMValCore/en/2189/recipe/preprocessor.html#volume-statistics
- https://esmvaltool--2189.org.readthedocs.build/projects/ESMValCore/en/2189/recipe/preprocessor.html#axis-statistics

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/preprocessor/_area.py
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_area.py:972]: do you want to catch a zero-division warning here and perform a check on the `area_stats_cube` if that pops up?

**Review by valeriupredoi — APPROVED** [review]
looks spiffy, bud :beer: One tiny comment from me

File: esmvalcore/preprocessor/_area.py
Comment (schlunma) [reply esmvalcore/preprocessor/_area.py:972]: Good point. I just tested this; with the current version of the code this already raises an `RuntimeWarning: invalid value encountered in divide` warning and gives `nan` values in the output. I can add a nicer message here. Would it also make sense to mask the cube instead of the `nan`s? I am not sure, maybe it also makes sense to keep the `nan`s to properly distinguish between the real mask and the zero-divisions.

Comment (schlunma) [reply esmvalcore/preprocessor/_area.py]: I found a weird problem with iris: https://github.com/SciTools/iris/issues/5523. Sometimes the output of zero divisions is masked, sometimes not. Thus, I chose to mask them consistently here.

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_area.py]: so in light of that issue, let's wait for the linked iris issue to be fixed before we do anything ourselves - in the lower stacks - it's fine to fix here (I reckon) :beers:

File: esmvalcore/_recipe/check.py
Comment (valeriupredoi) [inline esmvalcore/_recipe/check.py:489]: agh - Codacy may be right this time - I am a bit uneasy to mask something that's not TypeError nor ValueError and let it through just as a user warning - that may trigger a fall off the cliff anyway but harder to catch where from

File: esmvalcore/preprocessor/_area.py
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_area.py:976]:
```suggestion
        if (area_stats_cube.core_data() == 0.0).any():
```

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
nice warning! But area stats is the zero troublesome cube :smiley:

File: esmvalcore/_recipe/check.py
Comment (valeriupredoi) [reply esmvalcore/_recipe/check.py:489]: it won't terminate it here if it doesn't catch one of the (TypeError, ValueError) - I mean it will, but down the line when things will get stuck somewhere else - if, say, `check_preprocessor_settings(area_stats_settings_to_check)` fails with OSError (for some bizarre reason), the OSError will show up as a warning in the stdout ie "Caught OSError etc etc" but the thing will keep on chugging. Stuff will hit the fan later though

File: esmvalcore/preprocessor/_area.py
Comment (schlunma) [reply esmvalcore/preprocessor/_area.py:976]: Haha, true that!

**Review by valeriupredoi — APPROVED** [review]
cheers, Manu! Don't sweat too much about me comments about the RecipeError - all fine, and if the user doesn't know what went wrong, too bad, not my problem :rofl:

File: esmvalcore/_recipe/check.py
Comment (valeriupredoi) [reply esmvalcore/_recipe/check.py:489]: yes of course it will! I am a doofus, this is not in another try/except block of its own :man_facepalming:

**Comment by valeriupredoi** [comment]
@bouweandela would you be a darling a merge this, please? (I've been merging your stuff all morning :rofl: )

**Comment by bouweandela** [comment]
@schlunma Would it make sense to apply the same changes here as in #2191? I.e. remove the nesting from the function signature and it to:
```python
def normalize_with_area_statistics(
    cube: Cube,
    operation: str = 'subtract',
    operator: str,
    **kwargs: Any,
) -> Cube:
```
? (Or, if you think having bot operation and operator is confusing, rename one of the two?)

**Comment by schlunma** [comment]
I honestly think it is better the way it is. The keywords here are only relevant for the `area_statistics` part of this preprocessor, whereas in #2191 they are relevant for the actual preprocessors.

I have seen this kind of API in other packages as well, e.g., the `line_kws` in [`seaborn.regplot`](https://seaborn.pydata.org/generated/seaborn.regplot.html) which are only relevant for a part of the function.

**Comment by schlunma** [comment]
I would be nice to merge this PR after #2191; I want to add a proper check for the `area_statistics` keywords here introduced with #2191.

**Comment by bouweandela** [comment]
Thinking ahead a bit, this kind of feature could be implemented by adding just one extra keyword argument to all of the current `*_statistics` preprocessor functions. Maybe people would also be interested in knowing the spread of an ensemble for the `ensemble_statistics` preprocessor function, or normalizing by `volume_statistics`, etc. If we add new preprocessor functions for all of those we end up with many preprocessor functions. Are you really keen on getting this included into v2.10 (today is the deadline for the feature freeze) or would you be willing to think about it a bit more?

**Comment by schlunma** [comment]
I like the idea of generalizing this, but I don't think we should simply add a keyword to the existing `statistics` preprocessors. I think it would be weird if `area_statistics` somehow returned a cube that still has `latitude/longitude` dimensions.

It would probably be clearer to add a `normalize_with_statistics` preprocessor that takes the statistic as argument. Probably easy to implement for most preprocessors except for the multi-model ones since their call signature is different.  I can try to implement this for all preprocessors except for the multi-model ones today. Would that work for you?

**Comment by bouweandela** [comment]
That would probably be nicer than this approach, right?

**Comment by schlunma** [comment]
One issue with this generalized approach is that `normalize_with_statistics` requires different supplementary variables dependent on the statistic used (this is given as an argument to the function). Currently, the type of supplementary variable that a preperocessor uses is given with a [decorator](https://github.com/ESMValGroup/ESMValCore/blob/87e3f9fe95d2496222fb4029ce1e13feed1752b1/esmvalcore/preprocessor/_normalize.py#L18-L21), which knows nothing about the preprocessor arguments. In addition, if this `normalize_with_statistics` is used multiple times in the recipe, it might require different supplementary variables. This is also a problem for the current implementation, since the necessary supplementaries are [stored in a dictionary with the name of the preprocessor as key](https://github.com/ESMValGroup/ESMValCore/blob/87e3f9fe95d2496222fb4029ce1e13feed1752b1/esmvalcore/preprocessor/_supplementary_vars.py#L36).

@bouweandela any ideas how to solve this smartly without changing our entire code base?

**Comment by bouweandela** [comment]
In that case, adding the normalize option as an extra keyword argument to existing preprocessor functions may be a more attractive solution because it avoids all these technical issues. I agree that it may feel a bit counter-intuitive that the returned cube can have different dimensions based on the arguments though. Maybe we could ask the community for their opinion at the next monthly community meeting?

**Comment by bouweandela** [comment]
I just came across discussion https://github.com/ESMValGroup/ESMValTool/discussions/3353 where this seems relevant.

**Comment by schlunma** [comment]
> In that case, adding the normalize option as an extra keyword argument to existing preprocessor functions may be a more attractive solution because it avoids all these technical issues.

Agreed, that would be much simpler. I'll try to implement something like this for v2.11 because I need this new feature.

> I just came across discussion [ESMValGroup/ESMValTool#3353](https://github.com/ESMValGroup/ESMValTool/discussions/3353) where this seems relevant.

I don't think that's relevant if I understand the problem correctly (weighting != normalization).

File: doc/recipe/preprocessor.rst
Comment (valeriupredoi) [inline doc/recipe/preprocessor.rst:1929]:
```suggestion
      itself, but rather, the input cube, normalized with the statistics cube.
```

**Review by valeriupredoi — APPROVED** [review]
I believe I approved this PR once, in ancient times, not sure what happened in the meantime - still looks good to me, but I *think* we should dd a test for the volume normalization case too. At any rate, looking good, bar missing a few commas (yes, I am that kind of reviewer haha) :grin: :beer:

**Comment by valeriupredoi** [comment]
oh sorry, Manu - I got alerted I am up for providing a review - just saw it's still a draft - carry on, then

**Comment by schlunma** [comment]
This is ready for review now

**Review by valeriupredoi — APPROVED** [review]
this could be the fourth time I approve, but hey better more than never :grin:

**Comment by valeriupredoi** [comment]
not willing to approve this yet another time, I'll pass the ball to @bouweandela - if you are happy with all the stuff here, please merge, bud :beer:

**Comment by valeriupredoi** [comment]
frenly ping @bouweandela :baby_chick:

File: esmvalcore/preprocessor/_shared.py
Comment (bouweandela) [inline esmvalcore/preprocessor/_shared.py:221]: This kind of check can be computationally expensive if the cube is large. Instead of checking and warning at runtime, I would recommend just documenting the behaviour in the docstring (i.e. write that normalization with 'divide' will mask invalid values) and not do any checks at runtime because you're going to do this anyway and chances are real that users will not see the warning anyway in between the hundreds of warnings coming from iris.

File: doc/recipe/preprocessor.rst
Comment (bouweandela) [inline doc/recipe/preprocessor.rst:1957]: Would you be ok with calling this argument just `normalize` or `normalize_operator` instead? The abbreviation `stats` is for `statistics` is not immediately clear to me and writing `normalize_with_statistics` seems a bit long.

File: esmvalcore/preprocessor/_area.py
Comment (bouweandela) [inline esmvalcore/preprocessor/_area.py:240]: A bit of a nitpick, but I would find the following easier to read
```suggestion
    result = cube.collapsed('longitude', agg, **agg_kwargs)
    if normalize_with_stats is not None:
        result = get_normalized_cube(cube, result, normalize_with_stats)
```

**Review by bouweandela — COMMENTED** [review]
Thanks for making all the changes @schlunma! Looks great already, just a few more suggestions for making it even better. Feel free to merge as-is if you feel they're not relevant.

File: doc/recipe/preprocessor.rst
Comment (schlunma) [reply doc/recipe/preprocessor.rst:1957]: Still feels weird to me to call that simply `normalize`, but the current one is also certainly not optimal. Will change to `normalize`.

Comment (bouweandela) [reply doc/recipe/preprocessor.rst]: Thanks!

**Pull request merged**
